navigate to chat user's profile

rcullito opened the pull request:
Fixes #4279

Commit:
navigate to user profile from chat

rcullito commented:
A known issue is that this PR does not exactly fit the design spec here: zpl.io/V4J4NX4, mainly as we have not yet implemented dropdowns in RN for desktop. 

This PR addresses the issue of being able to navigate to the chat user's profile in the first place. Which can be done by clicking "View profile" in the upper right hand portion of the chat screen.

cammellos commented on src/status_im/chat/events.cljs:
```diff
@@ -400,13 +400,6 @@
                               (navigate-to-chat random-id {})
                               (transport.message/send (group-chat/GroupAdminUpdate. chat-name selected-contacts) random-id)))))
 
-(handlers/register-handler-fx
```
Looks like this is actually still used, why did you remove it?

rcullito replied:
@cammellos thanks for spotting. I think initially I didn't think this exact handler would be able to be reused for desktop. But an error on my part to remove it entirely. Have since added it back in.

cammellos commented on src/status_im/ui/components/desktop/events.cljs:
```diff
@@ -0,0 +1,16 @@
+    {:db (assoc db :desktop/render-user user)}))
+
+(handlers/register-handler-fx
+ :show-desktop-tab
```
How does the logic exactly works?

Why can't we use `current-account` directly instead of storing it in the database under a different key (`desktop/render-user`)?

rcullito replied:
The goal was to be able to re-use the same tab for displaying either the current user's information or the profile of the user the "current user" is chatting with. 

Previously the user profile tab subscribed to the current account, and I introduced this new key so that it could be more flexible. `desktop/render-user` can take on a value of either the current account or the person who is being chatted with. The different values are set based on whether the app user clicks on their own profile or someone else's. 

Ideally it would be nice to pass the account as an argument to the tab profile component, but given how it is currently structured, that struck me as a larger refactor beyond the scope of this change. Happy to make that as part of this or another PR if that seems more sensible in the long run.

cammellos replied:
@rcullito I see, makes sense, I would recommend to move away from denormalising and do as you proposed, as otherwise the view will not re-render if the subscription is cached and a contact update is received, unless you also update the rendered-user, which is probably something we want to avoid, and try to keep state normalized unless necessary.

This PR another PR is all fine by me :)

rcullito replied:
@cammellos Thanks for the review! Right, good point about the benefits of re-rendering. And sounds good, will aim to "re-normalize" in this PR then before it proceeds too much further. :+1:

cammellos commented on src/status_im/ui/screens/desktop/main/chat/views.cljs:
```diff
@@ -17,21 +17,25 @@
+         chat-name]
+                     (when (and (not group-chat) (not public?))
+                       [react/text {:on-press #(re-frame/dispatch [:show-desktop-tab :profile contact])}
+                       "View profile"])]
```
i18n

rcullito commented:
@EugeOrtiz your design feedback on this would be greatly appreciated in addition to our usual QA. While we currently don't have modal support, I had to make a decision about where to place the chat user's profile, and opted to re-use the component that we are already using for the logged in user's profile. Let me know if you'd like to talk any more about this!

churik commented:
Please, make sure that it is possible to build project on Jenkins.|
For now: 
https://github.com/status-im/status-react/pull/4893 cannot be built.
```
[MacOS parallel build stream] [31mWARNING[0m: Use of undeclared Var status-im.ui.screens.desktop.main.tabs.profile.views/profile at line 40 /Users/jenkins/workspace/sktop_pull-requests_PR-4893-OOJZB3WLBEPCUMJHRDGGECHNDWNBGRUFDQBP4NTHURV5TOWSMEWQ/src/status_im/ui/screens/desktop/main/tabs/profile/views.cljs
[MacOS parallel build stream] Error encountered performing task 'cljsbuild' with profile(s): 'prod'
[MacOS parallel build stream] Subprocess failed
```
https://jenkins.status.im/job/status-react/job/desktop/job/pull-requests/view/change-requests/job/PR-4893/28/console

<blockquote><img src="/static/f03d97b6/favicon.ico" width="48" align="right"><div><strong><a href="https://jenkins.status.im/job/status-react/job/desktop/job/pull-requests/view/change-requests/job/PR-4893/28/console">status-react » desktop » pull-requests » PR-4893 #28 Console [Jenkins]</a></strong></div></blockquote>

rcullito commented on src/status_im/ui/components/desktop/tabs.cljs:
```diff
@@ -38,9 +39,7 @@
   [react/touchable-highlight {:style    (merge tabs.styles/tab-container {:flex 1})
                               :disabled active?
-                              :on-press #(do
-                                           (re-frame/dispatch [:navigate-to :home])
```
@siphiuel think this line is removed from my merge with some of your code. lmk if my misunderstanding is off here and we actually do need this for proper navigation!

vitvly replied:
We don't want to persist right pane contents when navigating to a different tab. This was a problem with QR code PR, so I've added this event dispatch here. It resets right pane to a default view.

rcullito replied:
Ah ok, thank you clarifying this. And yes that makes total sense. Will add it back in. :+1:

rcullito commented:
Hey guys. Planning on adding functionality for "Add to contact" in a separate PR as it feels like that's a separate bit of functionality than just navigating to the user's profile. 

For testing purposes. The functionality contained in this one is seen by clicking "View Profile" in the upper right hand corner when chatting with someone. Then once their profile is open, it should display, image, name, contact code, and option to either return to chat or add contact. All of this will be there with the above note that "add contact" isn't wired up yet.

cammellos commented on src/status_im/ui/screens/desktop/main/chat/styles.cljs:
```diff
@@ -96,4 +94,38 @@
+(def contact-card-text
+  {:color     colors/blue
+   :font-size 14})
+
```
extra line

cammellos commented on src/status_im/ui/screens/desktop/main/tabs/profile/views.cljs:
```diff
@@ -32,6 +32,7 @@
        (i18n/label :sharing-copied-to-clipboard)]]
      [react/view {:style styles/tooltip-triangle}]]))
 
+
```
extra line

cammellos commented on src/status_im/ui/screens/desktop/main/views.cljs:
```diff
@@ -16,21 +16,22 @@
-                      status-view)]
-      [react/view {:style {:flex 1}}
-       [component]])))
+                 (let [component (case view-id
```
indentation

rcullito commented:
@cammellos thanks for spotting those!

vkjr commented on src/status_im/ui/screens/desktop/main/chat/views.cljs:
```diff
@@ -13,30 +13,36 @@
+  (views/letsubs
+   [{:keys [chat-id name public-key public? group-chat]} [:get-current-chat]
+    {:keys [pending?] :as contact}                       [:get-current-chat-contact]]
+   (let [chat-name (str
```
I would suggest extracting this to a separate function. That will make the code more readable and maybe it is not the only place where we can reuse it.

cammellos replied:
@vkjr good shout, actually there's already a subscription for that, `get-current-chat-name`, and should be used instead, reason being `:name` in chat is denormalized (needs to change everytime the contact updates her profile, therefore we are moving away from it)

rcullito replied:
good points! thanks both. will push a follow up change addressing this.

churik commented:
1) Not possible to navigate to user profile from Public chat (clicking on user picture)
2)  Icons are too small and different from the design
![icons](https://user-images.githubusercontent.com/4557972/42632888-654ada3c-85e7-11e8-9fdb-e1928b1eb061.png)
3) If User profile picture is white,edges are not visible (expected as grey lines, like in Mobile)
![phpto](https://user-images.githubusercontent.com/4557972/42632868-4f614f8a-85e7-11e8-82f0-419629b9de29.png)
4) No padding (should be 32 px according to design)
![padd](https://user-images.githubusercontent.com/4557972/42633115-1faf58e4-85e8-11e8-9842-8ffee3d560bb.png)
5) Contact code is not selectable for copying.
6) When User is in the Contact list, "In contacts" icon should be displayed. I believe even you are not implemented "Add to contact" functionality here, correct item have to be shown for already added user.
![incont](https://user-images.githubusercontent.com/4557972/42633765-1d33a26c-85ea-11e8-8515-fe997e48bcdf.png)

rcullito commented:
@churik thanks for your feedback! I've addressed most of your points, as well as those of @vkjr and @cammellos 

From our discussion last Friday, does it work if I make a separate PR for the remaining two items? This page re-uses a profile component from the other desktop tab, and I'm guessing viewing a profile of a user who is not already engaged in a chat with the current user will be a bit different in nature from what is in place currently. Still agreeing it needs to be done, just would like to tackle separately. Let me know if the changes checked off below seem valid. Thanks!

### Items to Address
- [x] remove chat-name code and use subscription 
- [x] update icons in chat profile (add is just a plus icon, for some reason other is not displaying well on desktop)
- [x] 32px padding for overall chat profile page
- [x] selectable contact code
- [x] dispay "in contacts" if the user is already in contacts
- [x] actually get "add to contacts" working from chat profile and have UI update accordingly 
- [ ] navigate to profile from public chat
- [ ] user profile photo gray border

churik commented:
7) Conflict with latest `desktop` branch (output from Jenkins is [here](https://jenkins.status.im/job/status-react/job/desktop/job/pull-requests/view/change-requests/job/PR-4893/56/console))
8) Expected, that contact code in Profile should be like selectable text (look to https://github.com/status-im/status-react/issues/5103 for details).
For now, it is not obvious that user copied a code if he just clicked on it - no messages, no warning, and text is not selecatable.

For item "Navigate to Profile from Public chat" - yes we can create separate PR, if you feel it more comfortable.
For grey borders I'll create a separate issue - this case is not handled through all app.

Commit:
go with develop styles for toolbar as that is a cleaner update

Commit:
working merge against develop. still need to fix right align for "view profile"

Commit:
fix alignment of view profile

Commit:
remove touchable-opacity from chat profile contact code and just make the text selectable

rcullito commented:
@churik thanks for those! k. 7 + 8 have been addressed when you have a chance to take another look at this. 

I'll open up another PR for navigating to profile from a public chat.

churik commented:
A separate issue for navigating to user profile from Public chat: 
https://github.com/status-im/status-react/issues/5184
The issue with transfer profile to popup: 
https://github.com/status-im/status-react/issues/5186

rcullito commented:
@churik great, thanks for testing and those separate issues. merging now :+1:
